server.js

```diff
@@ -60,21 +62,79 @@ passport.use('openidconnect', passportStrategy);
 passport.serializeUser((user, done) => done(null, user));
 passport.deserializeUser((user, done) => done(null, user));
 
+app.get('/', (req, res, next) => {
+  res.redirect('/hello');
+});
+
 // This routing path handles the start of an authentication request.
 // This is the path used in '/login.html' when you click the 'Sign in with Banno' button.
-app.get('/auth', passport.authenticate('openidconnect'));
+app.get('/auth', (req, res, next) => {
+  const options = {
+    // Random string for state
+    state: Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
+  };
+  req.session.oAuthState = req.session.oAuthState || {};
+  req.session.oAuthState[options.state] = {};
+  // If we have a deep link path query parameter, save it in a state parameter
+  // so that we can redirect to the correct page when the OAuth flow completes
+  // See https://auth0.com/docs/protocols/oauth2/redirect-users
+  if (req.query.returnPath && req.query.returnPath[0] === '/') {
+    req.session.oAuthState[options.state].returnPath = req.query.returnPath;
+  }
+  return passport.authenticate('openidconnect', options)(req, res, next);
+}
+);
 
 // This routing path handles the authentication callback.
 // This path (including the host information) must be configured in Banno SSO settings.
-app.get('/auth/cb', passport.authenticate('openidconnect', { successRedirect: '/me', failureRedirect: '/login' }));
+app.get('/auth/cb', (req, res, next) =>
+  passport.authenticate('openidconnect', (err, user, info) => {
+    console.log(err, user, info);
+    if (err || !user) {
+      return res.redirect('/login.html');
+    }
+    req.logIn(user, (err) => {
+      if (err) {
+        return next(err);
+      }
+      let nextPath = '/me';
+      // If a state parameter is present and has a matching local state, lookup the value
+      if (req.query.state) {
+        if (req.session.oAuthState && req.session.oAuthState[req.query.state]) {
+          if (req.session.oAuthState[req.query.state].returnPath) {
+            nextPath = req.session.oAuthState[req.query.state].returnPath;
+          }
+
+          delete req.session.oAuthState[req.query.state];
```

Comment:
_Non-blocking feedback:_

The repetition in using variable names `req.query.state` and `req.session.oAuthState` in tandem with nested `if` statements seems like it might need to be made a little bit easier to read.

With that said, I actually did try extracting these out into variables and broke the solution in my local environment. That's likely a result of me not being quite up to speed with modern JavaScript standards, but I think the feedback holds true.

For context, I'm an admittedly _irrational_ champion of the 🛡`guard` statement in Swift and might be trying to impose something that isn't idiomatic in _JavaScript_. Example from Swift: https://thatthinginswift.com/guard-statement-swift/

Reply:
There may be a slightly better way to do this in JS, but there isn't something like that guard statement in swift. Just a different language.